docs/design/2020-05-08-standardize-error-codes-and-messages.md

```diff
@@ -0,0 +1,208 @@
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+Check whether `tidb_disable_txn_auto_retry` is set to `on`. If so, set it to `off`; if it is already `off`, increase the value of `tidb_retry_limit` until the error no longer occurs.
+
+## Code: 9005
+### Error
+Region is unavailable
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+Check the status, monitoring data and log of the TiKV server.
+```
+
+Benefit:
+- It's easy to write.
+- It handles multiple lines well.
+- The markdown content can be previewed on writing.
+
+Tradeoff:
+- It's hard to implement (see example below)
+- It's not self-consistent (see example below)
+
+Tradeoff Example:
+
+```md
+## Code: 8005
+### Error
+Write Conflict, txnStartTS is stale
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+## Code: 9005
+### Error
+Region is unavailable
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+Check the status, monitoring data and log of the TiKV server.
+```
+
+As the syntax above, the 9005 block is the message part of 8005 block, so we expect it's result is the same as this toml:
+
+```toml
+[error.8005]
+error = '''Write Conflict, txnStartTS is stale'''
+message = '''Transactions in TiDB encounter write conflicts.'''
+workaround = '''
+## Code: 9005
+### Error
+Region is unavailable
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+Check the status, monitoring data and log of the TiKV server.
+'''
+```
+
+In this case, we must define the grammar and write a parser. Writing parser increase much complexity. What's worse
+is that I found no grammar that fits it.
+
+#### Summary
+
+Through the above discussion, we recommend the toml version of metafile. 
+
+In addition, the error codes should be append only in case of conflict between versions.
+
+### The Error Definition
+
+In the discussion above, an error has at least 4 parts:
+- The error code: it's the identity of an error
+- The error field: it's the error itself the user can view in TiDB system
+- the message field: the description of the error, what happened and why happend?
+- the workaround filed: how to workaround this error
+
+Besides, we can append a optional tags field to it:
+```toml
+[error.9005]
+error = ""
+message = ""
+workaround = ""
+tags = ["tikv"]
```

Comment:
do we need a standard of the "tag" field?

Reply:
The `tags` field is designed to classify errors (e.g. the level of seriousness). I think we can't classify them well before we have a list of them.
So I think the answer is yes, we need a standard, but not now. We can have it after we have a full list of the errors.

Reply:
OK. Could you add this explanation into the RFC itself? Rest LGTM.

Reply:
done